### PR TITLE
Adds BLS12 pairing function

### PR DESCRIPTION
**Pull request by armfazh**

### CODE REVIEW

File: ecc/bls12381/ff/cyclo12.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package ff
+
+import "fmt"
+
+// Cyclo12 represents an element of the 12-th cyclotomic group.
+type Cyclo12 [2]Fp6
+
+func (z Cyclo12) String() string           { return fmt.Sprintf("\n0: %v\n1: %v", z[0], z[1]) }
+func (z *Cyclo12) SetOne()                 { z[0].SetOne(); z[1].SetZero() }
```
Comment (wbl) [inline ecc/bls12381/ff/cyclo12.go:9]: This is the identity, and I've found it nice to have the generator as well, e.g e(g1, g2).

File: ecc/bls12381/ff/cyclo12.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
+package ff
+
+import "fmt"
+
+// Cyclo12 represents an element of the 12-th cyclotomic group.
+type Cyclo12 [2]Fp6
```
Comment (wbl) [inline ecc/bls12381/ff/cyclo12.go:6]: Eh, not when you expose Add and Sub. That makes it into a ring.

File: ecc/bls12381/ff/cyclo12.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+func (z Cyclo12) String() string           { return fmt.Sprintf("\n0: %v\n1: %v", z[0], z[1]) }
+func (z *Cyclo12) SetOne()                 { z[0].SetOne(); z[1].SetZero() }
+func (z *Cyclo12) Set(x *Cyclo12)          { z[0].Set(&x[0]); z[1].Set(&x[1]) }
+func (z *Cyclo12) Inv(x *Cyclo12)          { z.Set(x); z[1].Neg() }
+func (z *Cyclo12) IsZero() bool            { return z[0].IsZero() && z[1].IsZero() }
```
Comment (wbl) [inline ecc/bls12381/ff/cyclo12.go:12]: IsOne is really the thing I want.

File: ecc/bls12381/pair.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
+package bls12381
+
+import "github.com/cloudflare/circl/ecc/bls12381/ff"
+
+// Pair calculates the ate-pairing of P and Q.
+func Pair(P *G1, Q *G2) *Gt { return finalExp(miller(P, Q)) }
```
Comment (wbl) [inline ecc/bls12381/pair.go:6]: What I really want is products of pairings with exponents in them.

**Comment by wbl** [comment]
Looks pretty good, i left a few API comments.

File: ecc/bls12381/pair.go (hunk 4 above)
Comment (armfazh) [reply ecc/bls12381/pair.go:6]: This 



or

File: ecc/bls12381/ff/cyclo12.go (hunk 1 above)
Comment (armfazh) [reply ecc/bls12381/ff/cyclo12.go]: updated now

File: ecc/bls12381/ff/cyclo12.go (hunk 2 above)
Comment (armfazh) [reply ecc/bls12381/ff/cyclo12.go]: of course :)

File: ecc/bls12381/ff/cyclo12.go (hunk 3 above)
Comment (armfazh) [reply ecc/bls12381/ff/cyclo12.go]: done

File: ecc/bls12381/pair.go (hunk 4 above)
Comment (wbl) [reply ecc/bls12381/pair.go:6]: The former. We save time by integrating all those final exponents and the local ones, but doing it in constant time might be tricky. I'll see if I can extract something more specialized that would help.

**Review by wbl — APPROVED** [review]

File: ecc/bls12381/pair.go (hunk 4 above)
Comment (armfazh) [reply ecc/bls12381/pair.go:6]: Already included.

**Pull request merged**